Rewrite mania selection blueprint to use only `HitObject` (no `DrawableHitObject`) for layout

ekrctb opened the pull request:
- Fixes #13297
- [x] Depends on #13498

The selection blueprint test scene is rewritten.
I couldn't come up with a test reproducing the crash, because it is depending on one-frame window of column movement and snapping. The chance of regression should be low if `DrawableHitObject` is going to be eliminated from selection blueprint.

Also, It was not necessary to this PR, but  I removed the lifetime override logic of `DrawableManiaHitObject`. It was unnecessary (and conflicting) because DrawableHitObject now has its own logic for that.

ekrctb committed:
Use `HitObject` instead of DHO for mania selection blueprint layout

- Fix moving selected hold note between columns will cause a crash

ekrctb committed:
Introduce `IPlayfieldProvider`

ekrctb committed:
Rewrite mania selection blueprint test scene

ekrctb committed:
Add scrolling direction toggle for mania selection blueprint test scene

bdach committed:
Merge branch 'master' into fix-mania-editor-crash

bdach commented on osu.Game/Rulesets/Edit/IPlayfieldProvider.cs:
```diff
@@ -1,11 +1,12 @@
+namespace osu.Game.Rulesets.Edit
 {
-    public enum HoldNotePosition
+    public interface IPlayfieldProvider
```
I don't believe this interface is necessary to exist. An alternative way that I tested to work goes as follows:

* `ManiaSelectionBlueprintTestScene` is the easy case, as its `Playfield` can just have `[Cached(typeof(Playfield))]` on it.
* As for `HitObjectComposer<T>`, it's a bit trickier, but this can be done without the interface by overriding `CreateChildDependencies()`, creating a mutable dependency container there and storing it to a field, and then caching the `Playfield` in BDL. The [mania composer](https://github.com/ppy/osu/blob/044ff9d1d274cdc4ac1903c79064e97e3178c448/osu.Game.Rulesets.Mania/Edit/ManiaHitObjectComposer.cs#L85-L86) even does something like that already with `IScrollingInfo`.

Whether that's better, I'm not entirely sure, but it works without the interface and I'd personally prefer not to have it I think.

ekrctb replied:
I didn't know that way of using the dependency container. I appreciate.

bdach commented on osu.Game.Rulesets.Mania/Edit/Blueprints/ManiaSelectionBlueprint.cs:
```diff
@@ -29,19 +31,13 @@ protected override void Update()
 
-        public override void Show()
-        {
-            DrawableObject.AlwaysAlive = true;
```
Not sure why these are trimmed, but it seems like these were the only actual usages of this property. So it may be a good idea to remove them if they're no longer needed for some reason.

ekrctb replied:
It was unrelated to the changes of this PR, but the lifetime logic is no longer unnecessary (and conflicting) because `DrawableHitObject` has its own logic for that. I removed the lifetime logic.

bdach commented on osu.Game.Rulesets.Mania.Tests/Editor/ManiaSelectionBlueprintTestScene.cs:
```diff
@@ -1,31 +1,55 @@
     {
-        [Cached(Type = typeof(IAdjustableClock))]
-        private readonly IAdjustableClock clock = new StopwatchClock();
+        protected override Container<Drawable> Content => blueprints ?? base.Content;
```
What's the null fallback here for? `blueprints` can't possibly be null, can it?

ekrctb replied:
It is null when used during the base constructors.

bdach replied:
I would hope nobody is reading this fallback value, given that it may cause children to be added to a different container than expected, but that said other test scenes seem to be doing similar things so...

ekrctb committed:
Remove `IPlayfieldProvider` by caching `Playfield`

ekrctb committed:
Remove lifetime override of `DrawableManiaHitObject`

The `AlwaysAlive` logic is now in all DHOs and it is now not necessary (and potentially conflicting).

bdach reviewed (approved):
seems good

smoogipoo reviewed (approved):
lgtm!

smoogipoo committed:
Merge branch 'master' into fix-mania-editor-crash
